### PR TITLE
ARROW-7117: [C++][CI] Fix the hanging C++ tests in Windows 2019

### PR DESCRIPTION
**Pull request by pitrou**
The fix is simply to use normal shared linking.

### CODE REVIEW

File: .github/workflows/cpp.yml [hunk 1]
```diff
@@ -152,20 +152,20 @@ jobs:
...
-      ARROW_TEST_LINKAGE: static
+      ARROW_BUILD_SHARED: ON
+      ARROW_BUILD_STATIC: OFF
       ARROW_BOOST_USE_SHARED: OFF
-      ARROW_BUILD_SHARED: OFF
-      ARROW_USE_STATIC_CRT: ON
       ARROW_VERBOSE_THIRDPARTY_BUILD: OFF
     steps:
+      - name: Disable Crash Dialogs
+        run: reg add "HKCU\SOFTWARE\Microsoft\Windows\Windows Error Reporting" /v DontShowUI /t REG_DWORD /d 1 /f
```
Comment (pitrou) [inline .github/workflows/cpp.yml]: @kszucs Do you want me to put this in a script?

**Comment by pitrou** [comment]
Two questions:
1) can the decryption buffer grow unbounded?
2) do you have benchmark numbers?

File: .github/workflows/cpp.yml (hunk 1 above)
Comment (kszucs) [reply .github/workflows/cpp.yml]: This is quite environment specific so this is the right place.

Comment (pitrou) [reply .github/workflows/cpp.yml]: This can actually be useful on all Windows builders.

**Review by kszucs — APPROVED** [review]
LGTM

File: .github/workflows/cpp.yml (hunk 1 above)
Comment (pitrou) [reply .github/workflows/cpp.yml]: Note this is not required for this PR, but was part of my investigations and save future hassles.

**Comment by kszucs** [comment]
@pitrou Currently appveyor is the bottleneck for our overall CI times.

What do you think about removing https://github.com/apache/arrow/blob/master/appveyor.yml#L65-L68 ?

**Comment by pitrou** [comment]
I'd rather move it to GH Actions :-)

**Comment by pitrou** [comment]
AppVeyor: https://ci.appveyor.com/project/pitrou/arrow/builds/29155443

**Comment by kszucs** [comment]
My question is about how can we reach feature parity with that build.

**Comment by pitrou** [comment]
Ah. Well, regardless of feature parity, there are different dimensions being tested:
- different version of the MSVC compiler
- different CMake generators (msbuild vs Ninja)

So even if we enable features on the Win2019 build, we will still benefit from another build with Ninja and an older MSVC.

**Comment by kszucs** [comment]
Which MSVC version do we use on appveyor?

File: .github/workflows/cpp.yml (hunk 1 above)
Comment (kszucs) [reply .github/workflows/cpp.yml]: @pitrou then what was the solution? The hanging was just transient?

**Comment by kszucs** [comment]
The appveyor error is unrelated and probably transient, merging.

**Comment by pitrou** [comment]
Visual Studio 2015 on Windows Server 2012 is the oldest version, AFAIK.
Feel free to open a JIRA for creating new GH workflows. Also, remember that we're also testing Python on AppVeyor.

**Pull request closed without merging**